_posts/2021-09-29-user-defined-value-types-bug.md

```diff
@@ -0,0 +1,59 @@
+---
+layout: post
+published: true
+title: 'User Defined Value Types Bug'
+date: '2021-09-29'
+author: Solidity Team
+category: Security Alerts
+---
+
+On September 28th, 2021, Harry Altman ([@haltman-at](https://github.com/haltman-at)) of Truffle
+discovered a bug in user defined value types.
+
+The bug has no influence on the correctness of Solidity contracts, but contracts compiled with
+Solidity 0.8.8 that use the new feature are unnecessarily wasteful and might have problems with
+tooling or contract upgrades.
+
+The bug exists only in Solidity 0.8.8 and is fixed in 0.8.9.
+
+We assigned the bug a severity of "very very low".
+
+### Storage Layout of User Defined Value Types
+
+The compiler did not correctly compute the storage layout of user defined value types
+based on types that are shorter than 32 bytes. It would always use a full storage slot
+for these types, even if the underlying type is shorter.
+
+Types shorter than 32 bytes can share a storage slot with other values if there is enough space.
+
+For example:
+
+```solidity
+pragma solidity =0.8.8;
+
+type MyInt128 is int128;
+
+contract WastefulStorageLayout {
+    // slot 0
+    MyInt128 a;
+    // slot 1
+    MyInt128 b;
+}
+
+contract PackedStorageLayout {
+    // slot 0
+    int128 a;
+    // slot 0
+    int128 b;
+}
+```
+
+You can verify this by running `solc --storage-layout Contract.sol`.
+
+## Impact
+
+Solidity tries to guarantee that the storage layout of contracts remains consistent across releases.
+This allows for certain kind of upgradable patterns to exist that uses different Solidity versions.
+Since the Solidity 0.8.9 release changes the layout for short user defined value types, contract
+upgrades across versions that rely on consistent storage layouts may have bugs.
```

Comment:
Could we be more specific here? "Contract upgrades across 0.8.8 and 0.8.9 that rely on consistent storage layouts may introduce bugs."? Prior to user-defined-value type (introduced in 0.8.8) storage layout was consistent and it remains consistent for all types in 0.8.9 except UDVT right?

Reply:
I tried to me more specific now!